### PR TITLE
Dummy std lib

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
### What was wrong?

We don't have a standard library.

### How was it fixed?

A library ingot named `std` gets added to the global scope during compilation. It currently contains dummy code, which we'll replace once low-level intrinsics are supported.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/docs/src/spec/index.md) if applicable
- [x] Add entry to the [release notes](https://github.com/ethereum/fe/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

File: crates/test-files/fixtures/ingots/basic_ingot/src/bing.fe [hunk 1]
```diff
@@ -2,7 +2,7 @@ struct Bing:
     my_address: address
 
 fn get_42_backend() -> u256:
-    return 42
+    return std::get_42()
```
Comment (g-r-a-n-t) [inline crates/test-files/fixtures/ingots/basic_ingot/src/bing.fe:5]: check import error

File: crates/analyzer/tests/errors.rs [hunk 2]
```diff
@@ -45,12 +46,46 @@ fn error_string(path: &str, src: &str) -> String {
...
 
 fn error_string_ingot(path: &str) -> String {
-    let files = build_filestore(path);
+    let mut files = build_filestore(path);
 
     let db = TestDb::default();
 
-    let global = Global::default();
-    let global_id = db.intern_global(Rc::new(global));
+    let deps = indexmap! {
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/errors.rs:53]: try `Global::with_deps(...)`

File: crates/analyzer/src/db/queries/module.rs [hunk 3]
```diff
@@ -248,6 +254,14 @@ pub fn module_used_item_map(
                 diagnostics.extend(items.diagnostics.deref().clone());
 
                 for (name, (name_span, item)) in items.value.iter() {
+                    if module.global_ingots(db).contains_key(name) {
+                        diagnostics.push(errors::error(
+                            "a global ingot with the same name already exists",
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/db/queries/module.rs:259]: see how rust phrases this error and `use std`

File: crates/analyzer/src/namespace/items.rs [hunk 4]
```diff
@@ -204,13 +208,46 @@ impl Item {
...
 pub struct Global {
-    ingots: BTreeMap<String, IngotId>,
+    pub ingots: BTreeMap<String, IngotId>,
+}
+
+impl Global {
+    pub fn new(
+        db: &dyn AnalyzerDb,
+        files: &FileStore,
+        deps: &IndexMap<String, Vec<SourceFileId>>,
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:218]: handle dep compilation errors

File: crates/analyzer/src/namespace/items.rs [hunk 5]
```diff
@@ -221,10 +258,52 @@ pub struct Ingot {
     pub fe_files: BTreeMap<SourceFileId, (SourceFile, ast::Module)>,
 }
 
+impl Ingot {
+    pub fn new(
+        db: &dyn AnalyzerDb,
+        files: &FileStore,
+        name: &str,
+        file_ids: &[SourceFileId],
+        deps: &IndexMap<String, Vec<SourceFileId>>,
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:267]: handle dep compilation errors

File: crates/analyzer/src/namespace/items.rs [hunk 6]
```diff
@@ -221,10 +258,52 @@ pub struct Ingot {
...
+
+        Self {
+            name: name.to_string(),
+            global: global_id,
+            fe_files: file_ids
+                .iter()
+                .map(|file_id| {
+                    let file = files.get_file(*file_id).expect("missing file for ID");
+                    let (ast, parser_diagnostics) = parse_file(*file_id, &file.content)
+                        // .map_err(CompileError)
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:280]: use same return on moduleid

File: crates/analyzer/src/namespace/items.rs [hunk 7]
```diff
@@ -285,10 +368,62 @@ pub struct Module {
...
 
+impl Module {
+    pub fn new(
+        db: &dyn AnalyzerDb,
+        files: &FileStore,
+        file_id: SourceFileId,
+        deps: &IndexMap<String, Vec<SourceFileId>>,
+    ) -> Result<Analysis<Self>, Vec<Diagnostic>> {
+        let global = Global::new(db, files, deps);
+        let global_id = db.intern_global(Rc::new(global));
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:379]: use `GlobalId::new` and handle errors

File: crates/driver/src/lib.rs [hunk 8]
```diff
@@ -38,31 +35,13 @@ pub fn compile_module(
...
-            .expect("missing file name")
-            .to_string_lossy()
-            .to_string(),
-        context: ModuleContext::Global(global_id),
-        file_content: ModuleFileContent::File { file: file_id },
-        ast,
-    };
-    let module_id = db.intern_module(Rc::new(module));
+    let module_id = ModuleId::new(&db, files, file_id, &indexmap! {})
+        .unwrap()
```
Comment (g-r-a-n-t) [inline crates/driver/src/lib.rs:43]: handle error

File: crates/test-utils/src/lib.rs [hunk 9]
```diff
@@ -279,9 +280,14 @@ pub fn deploy_contract_from_ingot(
     contract_name: &str,
     init_params: &[ethabi::Token],
 ) -> ContractHarness {
-    let files = test_files::build_filestore(path);
+    let mut files = test_files::build_filestore(path);
+    let ingot_files = files.all_files();
 
-    let compiled_module = match driver::compile_ingot(path, &files, &files.all_files(), true, true)
+    let deps = indexmap! {
```
Comment (g-r-a-n-t) [inline crates/test-utils/src/lib.rs:286]: try to put this in one place

File: crates/analyzer/src/namespace/items.rs [hunk 10]
```diff
@@ -374,6 +548,24 @@ impl ModuleId {
...
     }
 
+    pub fn global(&self, db: &dyn AnalyzerDb) -> GlobalId {
+        match self.context(db) {
+            ModuleContext::Ingot(ingot) => ingot.data(db).global,
+            ModuleContext::Global(global) => global,
+        }
+    }
+
+    pub fn global_ingots(&self, db: &dyn AnalyzerDb) -> Rc<IndexMap<String, IngotId>> {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:558]: move into db

**Comment by g-r-a-n-t** [comment]
@sbillig lint is failing on some new conditions, but everything else is good, so feel free to take a look. Ill fix the clippy stuff before merging ]

**Comment by sbillig** [comment]
@g-r-a-n-t I'm trying to build on this, and running into trouble. If I add files to `library/std/src`, they get picked up in the `Dir` file list, but don't become child items of `std`.

Eg. added `library/std/src/evm.fe` and `library/std/src/foo/bar.fe`; `use std::evm::self_balance` and `use std::foo:bar::baz` fail with "unresolved path item" on `evm` and `foo` respectively

**Comment by g-r-a-n-t** [comment]
@sbillig huh, Ill look into it later today. I think it might be related to a change in the ingot_all_modules query

**Comment by sbillig** [comment]
```rust
    pub fn items(&self, db: &dyn AnalyzerDb) -> Rc<IndexMap<String, Item>> {
        match self {
            Item::Ingot(ingot) => ingot
                .lib_module(db)
                .expect("ingot is missing a lib file")
                .items(db),
```
A lib ingot is just the contents of lib.fe right now; the rest of the files are ignored.

**Comment by g-r-a-n-t** [comment]
> A lib ingot is just the contents of lib.fe right now; the rest of the files are ignored.

good catch. things are a bit tangled, but I think I know how to clean it up

**Comment by g-r-a-n-t** [comment]
Ok, I think items need to be teased apart into a few separate categories:

- **defined**: all code definitions (including sub modules)
- **global**: dep ingots and the prelude map
- **used**: items defined in use statements (looks inside of global and defined items)
- **internal**: items visible when inside of a module (defined, global, and used)
- **external**: items visible outside of the module (defined and used)

notes:
- when resolving a use path, we start by looking at global and defined items, then continue on looking at external items
- when resolving a type desc or expr, we start by looking at internal items, then continue looking at external items
- the point of having external items is to avoid things like `use my_module::address`

File: crates/analyzer/src/namespace/items.rs [hunk 11]
```diff
@@ -164,6 +142,7 @@ impl Item {
         match self.parent(db) {
             Some(Item::Ingot(_)) | None => Rc::new(vec![self.name(db)]),
             Some(parent) => {
+                // cycle here
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:145]: delete

File: crates/analyzer/src/namespace/items.rs [hunk 12]
```diff
@@ -180,6 +159,35 @@ impl Item {
         }
     }
 
+    pub fn resolve_path_segments(
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:162]: add doc comments to these methods

File: crates/analyzer/src/namespace/items.rs [hunk 13]
```diff
@@ -238,6 +384,22 @@ impl IngotId {
         db.ingot_main_module(*self).value
     }
 
+    pub fn lib_module(&self, db: &dyn AnalyzerDb) -> Option<ModuleId> {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:387]: comments on these

File: crates/analyzer/src/namespace/items.rs [hunk 14]
```diff
@@ -322,42 +533,218 @@ impl ModuleId {
...
+        sub_modules
+            .deref()
+            .clone()
+            .into_iter()
+            .map(|(name, module)| (name, Item::Module(module)))
+            .chain(global_items)
+            .collect()
+    }
+
+    pub fn internal_items(&self, db: &dyn AnalyzerDb) -> IndexMap<SmolStr, Item> {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:564]: docs

File: crates/analyzer/src/namespace/items.rs [hunk 15]
```diff
@@ -322,42 +533,218 @@ impl ModuleId {
...
         &self,
         db: &dyn AnalyzerDb,
         tree: &Node<ast::UseTree>,
-    ) -> Rc<IndexMap<SmolStr, (Span, Item)>> {
-        db.module_resolve_use_tree(*self, tree.to_owned()).value
+        internal: bool,
+    ) -> Analysis<Rc<IndexMap<SmolStr, (Span, Item)>>> {
+        let mut diagnostics = vec![];
+
+        let resolve_path = |module: &ModuleId, db: &dyn AnalyzerDb, path: &ast::Path| {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:635]: comment explaining

File: crates/analyzer/src/namespace/items.rs [hunk 16]
```diff
@@ -377,6 +764,31 @@ impl ModuleId {
...
     }
 
+    pub fn global(&self, db: &dyn AnalyzerDb) -> GlobalId {
+        match self.context(db) {
+            ModuleContext::Ingot(ingot) => ingot.data(db).global,
+            ModuleContext::Global(global) => global,
+        }
+    }
+
+    pub fn global_items(&self, db: &dyn AnalyzerDb) -> IndexMap<SmolStr, Item> {
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/items.rs:774]: docs

File: crates/analyzer/src/namespace/scopes.rs [hunk 17]
```diff
@@ -194,7 +194,10 @@ impl<'a> AnalyzerContext for FunctionScope<'a> {
     }
 
     fn resolve_path(&mut self, path: &ast::Path) -> Option<NamedThing> {
-        let item = Item::Module(self.function.module(self.db())).resolve_path(self.db(), path);
+        let item = self
+            .function
+            .module(self.db())
+            .resolve_path(self.db(), path);
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/namespace/scopes.rs:200]: this should probably be `resolve_path_internal`

File: crates/analyzer/src/traversal/expressions.rs [hunk 18]
```diff
@@ -891,6 +891,7 @@ fn expr_call_path<T: std::fmt::Display>(
     generic_args: &Option<Node<Vec<fe::GenericArg>>>,
     args: &Node<Vec<Node<fe::CallArg>>>,
 ) -> Result<(ExpressionAttributes, CallType), FatalError> {
+    println!("here");
```
Comment (g-r-a-n-t) [inline crates/analyzer/src/traversal/expressions.rs:894]: delete

File: crates/analyzer/tests/analysis.rs [hunk 19]
```diff
@@ -74,39 +75,21 @@ macro_rules! test_analysis_ingot {
         #[test]
         #[wasm_bindgen_test]
         fn $name() {
-            let files = test_files::build_filestore($path);
+            // fix this
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/analysis.rs:78]: fix what?

File: crates/analyzer/tests/analysis.rs [hunk 20]
```diff
@@ -74,39 +75,21 @@ macro_rules! test_analysis_ingot {
...
-                        };
-                        (file.id, (file.clone(), ast))
-                    })
-                    .collect(),
-            };
-            let ingot_id = db.intern_ingot(Rc::new(ingot));
+            let ingot_id = analysis.value;
+
+            if !analysis.diagnostics.deref().is_empty() {
+                panic!("")
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/analysis.rs:91]: print errors

File: crates/analyzer/tests/errors.rs [hunk 21]
```diff
@@ -1,42 +1,23 @@
...
-
-    let module = items::Module {
-        name: path.into(),
-        context: items::ModuleContext::Global(global_id),
-        file_content: ModuleFileContent::File { file: id },
-        ast,
-    };
-
-    let module_id = db.intern_module(Rc::new(module));
+    let module_id = ModuleId::try_new(&db, &files, id, &indexmap! {})
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/errors.rs:18]: this needs deps

File: crates/analyzer/tests/snapshots/errors__bad_ingot.snap [hunk 22]
```diff
@@ -3,18 +3,6 @@ source: crates/analyzer/tests/errors.rs
 expression: error_string_ingot(&path)
 
 ---
-error: cannot glob import from type
```
Comment (g-r-a-n-t) [inline crates/analyzer/tests/snapshots/errors__bad_ingot.snap:6]: error?

**Comment by g-r-a-n-t** [comment]
Alright @sbillig, this is ready for another review. I basically did what was described in my previous comment and fixed the issue you initially pointed out.

As mentioned before, I didn't overhaul the item map queries yet, so it's possible to have cycles on `use` paths. I don't think this a blocker for real `std` lib stuff, tho.

**Comment by g-r-a-n-t** [comment]
For clarity, cycles occur when two modules contain use paths that point at each other.
e.g.
```
foo.fe
---
use ingot::bar::... # must resolve all items in bar

bar.fe
---
use ingot::foo::... # must resolve all items in foo. cycles.
```

**Review by sbillig — APPROVED** [review]
:man_cook: :+1:

**Pull request merged**